tensorflow-frameworks/tensorflow-data/src/main/java/org/tensorflow/data/Dataset.java

```diff
@@ -0,0 +1,180 @@
+  public final Dataset batch(long batchSize) {
+    return batch(batchSize, true);
+  }
+
+  /**
+   * Creates new `Dataset` skips `count` initial elements from this dataset
+   *
+   * @param count The number of elements to `skip` to form the new dataset.
+   * @return A new Dataset with `count` elements removed.
+   */
+  public final Dataset skip(long count) {
+    return new SkipDataset(tf, this.getVariant(), tf.val(count), this.getOutputTypes(), this.getOutputShapes());
+  }
+
+  /**
+   * Creates new `Dataset` with the first `count` elements from this dataset.
+   *
+   * @param count The number of elements to "take" from this dataset.
+   * @return A new Dataset containing the first `count` elements from this dataset.
+   */
+  public final Dataset take(long count) {
+    return new TakeDataset(tf, this.getVariant(), tf.val(count), this.getOutputTypes(), this.getOutputShapes());
+  }
+
+  /**
+   * Creates an iterator which iterates through all batches of this Dataset in an eager fashion.
+   * Each batch is a list of components, returned as `Output` objects.
+   * <p>
+   * This method enables for-each iteration through batches when running
+   * in eager mode. For Graph mode batch iteration, see `makeOneShotIterator`.
+   *
+   * @return an Iterator through batches of this dataset.
+   */
+  @Override
+  public Iterator<List<Output<?>>> iterator() {
+
+    if (!(tf.scope().env() instanceof EagerSession)) {
+      throw new UnsupportedOperationException("Cannot iterate through a dataset in graph mode.");
+    }
+
+    Operand<?> dataset = getVariant();
+    AnonymousIterator anonymousIterator = tf.data.anonymousIterator(getOutputTypes(), getOutputShapes());
+
+    tf.data.makeIterator(dataset, anonymousIterator.handle());
+
+    return new Iterator<List<Output<?>>>() {
+      private List<Output<?>> tryNext = getNext();
+
+      private List<Output<?>> getNext() {
+        try {
+          return tf.data.iteratorGetNext(anonymousIterator.handle(), getOutputTypes(), getOutputShapes()).components();
+        } catch (IndexOutOfBoundsException e) {
+          return null;
+        }
+      }
+
+      @Override
+      public boolean hasNext() {
+        return tryNext != null;
+      }
+
+      @Override
+      public List<Output<?>> next() {
+        List<Output<?>> result = tryNext;
+        tryNext = getNext();
+        return result;
+      }
+    };
+  }
+
+  /**
+   * Return the necessary components to iterate through batches of this
+   * dataset in Graph mode.
+   * <p>
+   * This method returns a Pair whose first element is a MakeIterator operation
+   * that must be run first in its own session to create the iterator internally.
+   * <p>
+   * The second element in the pair is a list of Output objects. In sequential
+   * calls to session.run() in which these (or child) nodes are fetched, the batches
+   * are already loaded into these objects.
+   *
+   * @return A Pair whose first element is a MakeIterator Operation, and whose

```

Comment:
It seems that it does not return a `Pair` anymore?

Reply:
Good catch, yup this returns `OneShotIterator` now.

Reply:
... And now usees `DatasetIterator`